src/query/api/v1/handler/prometheus/native/read_common.go

```diff
@@ -204,29 +199,35 @@ func sortedBlocksToSeriesList(blockList []blockWithMeta) ([]*ts.Series, error) {
 		numValues += b.StepCount()
 	}
 
-	for i := 0; i < numSeries; i++ {
-		values := ts.NewFixedStepValues(bounds.StepSize, numValues,
-			math.NaN(), bounds.Start)
-		valIdx := 0
-		for idx, iter := range seriesIters {
-			if !iter.Next() {
-				if err = iter.Err(); err != nil {
-					return nil, err
-				}
-
-				return nil, fmt.Errorf(
-					"invalid number of datapoints for series: %d, block: %d", i, idx)
-			}
+	// TODO: fix this to not alloc a big block every time.
+	data := make([][]float64, numSeries)
+	valuesBatchAlloc := make([]float64, numSeries*numValues)
+	for i := range data {
+		data[i] = valuesBatchAlloc[:numValues]
+		valuesBatchAlloc = valuesBatchAlloc[numValues:]
+	}
 
-			if err = iter.Err(); err != nil {
-				return nil, err
+	rowCount := 0
+	for _, it := range iters {
+		for it.Next() {
+			step := it.Current()
+			for i, v := range step.Values() {
+				data[i][rowCount] = v
```

Comment:
Just curious, why so necessary to save this into `data/values` now instead of just the values with SetValueAt like previously?

Reply:
Sound good, made the change like we discussed 👍